core/src/main/java/org/apache/druid/data/input/IntermediateRowParsingReader.java

```diff
@@ -93,51 +111,126 @@ public InputRow next()
+                    source(),
+                    null,
+                    metadata
+                ))
+            );
+          }
 
-      List<InputRow> rows;
-      try {
-        rows = parseInputRows(row);
-      }
-      catch (ParseException e) {
-        return InputRowListPlusRawValues.ofList(rawColumnsList, e);
+          if (CollectionUtils.isNullOrEmpty(rawColumnsList)) {
+            return InputRowListPlusRawValues.of(
+                null,
+                new ParseException(String.valueOf(row), buildParseExceptionMessage(
+                    StringUtils.nonStrictFormat("No map object parsed for row [%s]", row),
+                    source(),
+                    null,
+                    metadata
+                ))
+            );
+          }
+
+          List<InputRow> rows;
+          try {
+            rows = parseInputRows(row);
+          }
+          catch (ParseException e) {
+            return InputRowListPlusRawValues.ofList(rawColumnsList, new ParseException(
+                String.valueOf(row),
+                e,
+                buildParseExceptionMessage(e.getMessage(), source(), null, metadata)
+            ));
+          }
+          catch (IOException e) {
+            ParseException exception = new ParseException(String.valueOf(row), e, buildParseExceptionMessage(
+                StringUtils.nonStrictFormat("Unable to parse row [%s] into inputRow", row),
+                source(),
+                null,
+                metadata
+            ));
+            return InputRowListPlusRawValues.ofList(rawColumnsList, exception);
+          }
+
+          return InputRowListPlusRawValues.ofList(rawColumnsList, rows);
+        };
+
+    return new CloseableIterator<InputRowListPlusRawValues>()
+    {
+      @Override
+      public void close() throws IOException
+      {
+        delegate.close();
       }
-      catch (IOException e) {
-        ParseException exception = new ParseException(String.valueOf(row), e, "Unable to parse row [%s] into inputRow", row);
-        return InputRowListPlusRawValues.ofList(rawColumnsList, exception);
+
+      @Override
+      public boolean hasNext()
+      {
+        return delegate.hasNext();
       }
 
-      return InputRowListPlusRawValues.ofList(rawColumnsList, rows);
-    });
+      @Override
+      public InputRowListPlusRawValues next()
+      {
+        if (!hasNext()) {
+          throw new NoSuchElementException();
+        }
+
+        return samplingFunction.apply(delegate.next(), delegate.metadata());
+      }
+    };
   }
 
   /**
    * Creates an iterator of intermediate rows. The returned rows will be consumed by {@link #parseInputRows} and
-   * {@link #toMap}.
+   * {@link #toMap}. Either this or {@link #intermediateRowIteratorWithMetadata()} should be implemented
+   */
+  protected CloseableIterator<T> intermediateRowIterator() throws IOException
+  {
+    throw new UnsupportedEncodingException("intermediateRowIterator not implemented");
```

Comment:
UnsupportedEncodingException doesn't look right. It should be UnsupportedOperationException here.

Reply:
Added accidentally